watertap/unit_models/leaky_pressure_exchanger.py

```diff
@@ -0,0 +1,491 @@
+        self.low_pressure_side.add_momentum_balances(
+            balance_type=self.config.momentum_balance_type,
+            has_pressure_change=True)
+
+        self.low_pressure_side.deltaP.setlb(0)
+
+        @self.low_pressure_side.Expression(
+            self.flowsheet().config.time,
+            doc='Work transferred to low pressure side fluid')
+        def work(b, t):
+            return b.properties_in[t].flow_vol * b.deltaP[t]
+
+        @self.low_pressure_side.Expression(
+            self.flowsheet().config.time,
+            self.config.property_package.phase_list,
+            self.config.property_package.component_list,
+            doc='Mass transfered on low pressure side (will be negative, gained mass)')
+        def mass_transfer(b, t, p, j):
+            return b.properties_in[t].flow_mass_phase_comp[p,j]-b.properties_out[t].flow_mass_phase_comp[p,j]
+
+        #Add Ports
+        self.add_inlet_port(name='high_pressure_inlet', block=self.high_pressure_side)
+        self.add_outlet_port(name='high_pressure_outlet', block=self.high_pressure_side)
+        self.add_inlet_port(name='low_pressure_inlet', block=self.low_pressure_side)
+        self.add_outlet_port(name='low_pressure_outlet', block=self.low_pressure_side)
+
+        # Performance equations
+        @self.Constraint(
+            self.flowsheet().config.time,
+            doc="Pressure transfer")
+        def eq_pressure_transfer(b, t):
+            return (b.low_pressure_side.deltaP[t] ==
+                    b.efficiency_pressure_exchanger[t] * -b.high_pressure_side.deltaP[t])
+
+        @self.Constraint(
+            self.flowsheet().config.time,
+            doc="Equal volumetric flow rate")
+        def eq_equal_flow_vol(b, t):
+            return (b.high_pressure_side.properties_in[t].flow_vol ==
+                    b.low_pressure_side.properties_in[t].flow_vol)
+
+        @self.Constraint(
+            self.flowsheet().config.time,
+            self.config.property_package.phase_list,
+            self.config.property_package.component_list,
+            doc="mass leak from HP inlet to LP outlet")
+        def eq_mass_transfer_from_high_to_low_pressure_outlet(b, t, p, j):
+            comp = self.config.property_package.get_component(j)
+            if comp.is_solvent():
+                return (b.low_pressure_side.properties_out[t].flow_mass_phase_comp[p,j]==\
+                b.high_pressure_side.properties_in[t].flow_mass_phase_comp[p,j]*b.solvent_leakage_fraction[t]+
+                b.low_pressure_side.properties_in[t].flow_mass_phase_comp[p,j])
+            if comp.is_solute():
+                return (b.low_pressure_side.properties_out[t].flow_mass_phase_comp[p,j]==\
+                b.high_pressure_side.properties_in[t].flow_mass_phase_comp[p,j]*b.solute_leakage_fraction[t]+
+                b.low_pressure_side.properties_in[t].flow_mass_phase_comp[p,j])
+
+        ### this constraint (eq_mass_balance_high_pressure_outlet) might not be needed###
```

Comment:
This is definitely not needed - it is redundant when combined with the mass balances in the control volumes and the above constraint, If it is not removed, then the model will be degenerate and will not solve.

Reply:
Hmm, it does solve with this, but will remove.

Reply:
Surprisingly, it solves without errors in a full RO model. Should we be expecting errors to pop up, or was I just lucky? Will remove in next version.

Reply:
I would expect to see two things:

1. There will be a degree of freedom issue somewhere. This appears as an extra constraint in the problem, so when aiming to get 0 degrees of freedom this means that we will be able to fix one less variable.
2. The problem with degeneracies is that there still exist feasible solutions, and with good initial guesses you can end up with one of these. However, the Jacobian of the problem is singular, which means if you need to find a new feasible point, some of the information is missing. I.e. degeneracies can often go unnoticed as they can be solved under specific conditions, but can suddenly start failing when you try to do something else.

Reply:
That makes sense. Thanks for the info.